homeassistant/components/bluesound/coordinator.py

```diff
@@ -0,0 +1,178 @@
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+NODE_OFFLINE_CHECK_TIMEOUT = timedelta(minutes=3)
+PRESET_AND_INPUTS_INTERVAL = timedelta(minutes=15)
+
+
+@dataclass
+class BluesoundData:
+    """Define a class to hold Bluesound data."""
+
+    sync_status: SyncStatus
+    status: Status
+    presets: list[Preset]
+    inputs: list[Input]
+
+    is_online: bool
+
+
+def cancel_task(task: asyncio.Task) -> Callable[[], Coroutine[None, None, None]]:
+    """Cancel a task."""
+
+    async def _cancel_task() -> None:
+        task.cancel()
+        with contextlib.suppress(asyncio.CancelledError):
+            await task
+
+    return _cancel_task
+
+
+class BluesoundCoordinator(DataUpdateCoordinator[BluesoundData]):
+    """Define an object to hold Bluesound data."""
+
+    def __init__(
+        self, hass: HomeAssistant, player: Player, sync_status: SyncStatus
+    ) -> None:
+        """Initialize."""
+        self.player = player
+        self._inital_sync_status = sync_status
+
+        super().__init__(
+            hass,
+            logger=_LOGGER,
+            name=sync_status.name,
+        )
+
+    async def _async_setup(self) -> None:
+        assert self.config_entry is not None
+
+        preset = await self.player.presets()
+        inputs = await self.player.inputs()
+        status = await self.player.status()
+
+        self.async_set_updated_data(
+            BluesoundData(
+                sync_status=self._inital_sync_status,
+                status=status,
+                presets=preset,
+                inputs=inputs,
+                is_online=True,
+            )
+        )
+
+        status_loop_task = self.hass.async_create_background_task(
+            self._poll_status_loop(),
+            name=f"bluesound.poll_status_loop_{self.data.sync_status.id}",
+        )
+        self.config_entry.async_on_unload(cancel_task(status_loop_task))
+
+        sync_status_loop_task = self.hass.async_create_background_task(
+            self._poll_sync_status_loop(),
+            name=f"bluesound.poll_sync_status_loop_{self.data.sync_status.id}",
+        )
+        self.config_entry.async_on_unload(cancel_task(sync_status_loop_task))
+
+        presets_and_inputs_loop_task = self.hass.async_create_background_task(
+            self._poll_presets_and_inputs_loop(),
+            name=f"bluesound.poll_presets_and_inputs_loop_{self.data.sync_status.id}",
+        )
+        self.config_entry.async_on_unload(cancel_task(presets_and_inputs_loop_task))
+
+    async def _async_update_data(self) -> BluesoundData:
+        return self.data
+
+    def _set_offline(self) -> None:
+        _LOGGER.error(
+            "Node %s is offline or threw an unexpected error, retrying later",
+            self.data.sync_status.id,
+        )
+        self.async_set_updated_data(
+            replace(
+                self.data,
+                is_online=False,
```

Comment:
Hmm, I am not sure if I like the `is_online` attribute over using `async_set_update_error`. The latter has the added bonus that you don't need extra code in the entities for handling being unavailable and it handles logging for you

Reply:
I switched to using `async_set_update_error`.